## Reinforcement Learning
#### Reinforcement Learning is a common Machine Learning techniques where the model learns through trail and error. Reinforcement Learning is commonly used for video games, and robotics. The common features of RL are listed below:
- Agent: Learner who decides where to move
- Policy: A rule that guides the decision process of the agent
- Environment: The area in where all possible states are
- State: Current state of agent
- Action: The act of moving from one state to another
- Goal State: the state which agent is looking for. This is the state with reward.

## Robot Game
#### In this game we will focus on a robot who will try to make all rooms(states); A, B, C, D clean. The environemnt consists of these four rooms, and will be reffered to as states. Using the policy_evaluation function, we will find the state values, s1, s2, s3, s4. Rewards of each actios are listed below:
- Each movement between the rooms has a reward of −0.04.
- Suck action has a reward of −0.1.
- Reaching the terminal state s5 will lead to a reward of 1.0.
- The discount factor is γ = 0.9.

In [1]:
import numpy as np

# The robot aims to clean the dusts in room D. The layout of the rooms are shown below.
# --------------
# |  A  |  B   |
# --------------
# |  C  | .D.  |
# --------------

# Define states
states = [
    's1',  # at A, D uncleaned
    's2',  # at B, D uncleaned
    's3',  # at C, D uncleaned
    's4',  # at D, D uncleaned
    's5'   # at D, D cleaned
]

# Terminal state is s5, where D is cleaned
terminal_states = ['s5']

# Define actions
actions = ['left', 'right', 'up', 'down', 'suck']

# Applicable actions for each state, except terminal states
applicable_actions = [
    ['right', 'down', 'suck'],  # for state s1
    ['left', 'down', 'suck'],  # for state s2
    ['right', 'up', 'suck'],  # for state s3
    ['left', 'up', 'suck']  # for state s4
]

# Next state when applying each applicable action in each state
next_states = [
    ['s2', 's3', 's1'],   # s'(s1, a)
    ['s1', 's4', 's2'],   # s'(s2, a)
    ['s4', 's1', 's3'],   # s'(s3, a)
    ['s3', 's2', 's5']    # s'(s4, a)
]

# Reward when applying each applicable action in each state
rewards = [
    [-0.04, -0.04, -0.1],   # r(s1, a)
    [-0.04, -0.04, -0.1],   # r(s2, a)
    [-0.04, -0.04, -0.1],   # r(s3, a)
    [-0.04, -0.04, -0.1]    # r(s4, a)
]

# Discount factor
gamma = 0.9

# When reaching the terminal state s5, we receive a reward of 1.0.
terminal_reward = 1.0

###

In [12]:
# Q1.1: policy evaluation

def policy_evaluation(policy):
    V = {}
    for s in states:
        if s in terminal_states:
            V[s] = terminal_reward
        else:
            V[s] = 0.0
        
    while True:
        delta = 0
        for i, s in enumerate(states[:-1]):
            vold = V[s]
            action = policy[s]
            a_idx = applicable_actions[i].index(action)
            next_state = next_states[i][a_idx]
            reward = rewards[i][a_idx]

            V[s] = reward + gamma * V[next_state]

            delta = max(delta, abs(vold - V[s]))

        if delta == 0:
            return [V[s] for s in states[:-1]]
    


pi = {
        's1': 'right',
        's2': 'suck',
        's3': 'up',
        's4': 'up'}

state_values = policy_evaluation(pi)
print(state_values)

[-0.9399999999999995, -0.9999999999999994, -0.8859999999999996, -0.9399999999999995]


## Results
#### We recieved the state values, [s1= -0.93, s2= -1, s3= -0.89, s4= -0.93]. These values are highly negative. Which shows us that starting in any of these states following the policy will give us bad outcomes. The robot should not follow this policy.

In [92]:
###
# Q1.2: value iteration
###

# @TODO
# Please complete this value iteration method,
# which returns the optimal state values for all the states,
# except the terminal states.
# Please delete the "return 0;" as it is a placeholder.
def value_iteration():
    V = {}
    for s in states:
        if s in terminal_states:
            V[s] = terminal_reward
        else:
            V[s] = 0.0
        
    while True:
        delta = 0
        qvalue = {}
        for i, s in enumerate(states[:-1]):
            vold = V[s]
            V[s] = 0.8 * (rewards[i][a_idx] + gamma0* V[next_states[i][a_idx]] 
                       for a_idx in range(len(applicable_actions[i])))
                
                

            delta = max(delta, abs(vold - V[s]))

        if delta == 0:
            return [V[s] for s in states[:-1]]


# Run value iteration
state_values = value_iteration()
print(state_values)

[0.5720000000000001, 0.68, 0.68, 0.8]


## Results
- The state values we recived [s1= 0.58, s2= 0.68, s3= 0.68, s4= 0.8], these values differ from previous policy. This is because in this policy, we iterate to find the optimal soluton for each state. The highest value we recieved was from s4, which was a state value of 0.8. This is a high positive value and shows us that this state benefits highly from its action. This is expected because the robot is trying to clean the dirt which is only found in s4. We recieved slightly lower values in state 2 and state 3, because it would take one step before reaching the goal state which is state 4. The lowest state value is recieved from state 1, becuase it would require two steps before reaching goal state, requiring extra costs. 
- This policy(value iteration) follows a deterministic structure and found the optimal state value for these states by updating the costs until there are no changes. This model considered all possible moves at each state and returned state 4 with highest reward.